README.md
# Introduction

A basic library for numerical analysis in JavaScript, mostly for my 
own enjoyment.

I tried writing it for efficiency, but also style. It's unit tested 
and commented.

For offline development, I have been using `Node.js`.

# Dependencies

Unit testing is handled with [Jasmine](http://jasmine.github.io/), 
mostly offline...so with 
[jasmine-node](https://github.com/mhevery/jasmine-node) to be precise.

# License

This is released under the MIT license.


changelog.md
# Initial Relase
- Used `significand` instead of `mantissa`, following Knuth
- Followed Java's helper functions for `util.scalb()` and `DoubleConsts`

- TODO:
  - Need to make sure the bit-fiddling works around endianness


spec/util-spec.js
var util = require('../src/util');
var DoubleConsts = require('../src/DoubleConsts');

var Gelfond = Math.pow(Math.E, Math.PI);
var MAX_SCALE = DoubleConsts.MAX_EXPONENT + -DoubleConsts.MIN_EXPONENT
                + DoubleConsts.SIGNIFICAND_WIDTH + 1;
var Double_MAX_VALUEmm = "1.7976931348623155E308";
var Double_MAX_SUBNORMAL = "2.225073858507201E-308";
var Double_MAX_SUBNORMALmm = "2.2250738585072004E-308";

describe("getNumberParts() tests", function() {
    var checkNumber = function(x) {
        var parts = util.getNumberParts(x);
        return Math.pow(-1, parts.sign)*Math.pow(2, parts.exponent)*parts.significand;
    }
    it("should handle one", function () {
        expect(1).toBe(checkNumber(1));
    });
    it("should handle negative numbers", function() {
        expect(-1).toBe(checkNumber(-1));
    });
    it("should handle transcendental numbers", function() {
        expect(Math.PI).toBe(checkNumber(Math.PI));
        expect(Math.E).toBe(checkNumber(Math.E));
        expect(Gelfond).toBe(checkNumber(Gelfond));
    });
    it("should handle infinities", function() {
        expect(Infinity).toBe(checkNumber(Infinity));
        expect(-Infinity).toBe(checkNumber(-Infinity));
    });
});

describe("fromNumberParts() tests", function() {
    var checkNumber = function(x) {
        var parts = util.getNumberParts(x);
        return util.fromNumberParts(parts.sign, parts.significand, parts.exponent);
    }
    it("should handle one", function () {
        expect(1).toBe(checkNumber(1));
    });
    it("should handle negative numbers", function() {
        expect(-1).toBe(checkNumber(-1));
    });
    it("should handle transcendental numbers", function() {
        expect(Math.PI).toBe(checkNumber(Math.PI));
        expect(Math.E).toBe(checkNumber(Math.E));
        expect(Gelfond).toBe(checkNumber(Gelfond));
    });
    it("should handle infinities", function() {
        expect(Infinity).toBe(checkNumber(Infinity));
        expect(-Infinity).toBe(checkNumber(-Infinity));
    });
});

describe("scalb() function", function() {
    function scalbTest(value, scaleFactor, expected) {
        expect(util.scalb(value, scaleFactor)).toBe(expected);
        expect(util.scalb(-value, scaleFactor)).toBe(-expected);
    }
    var identityCases = [-0.0, +0.0, Infinity];
    var subnormalCases = [Number.MIN_VALUE,
                          3.0*Number.MIN_VALUE,
                          Double_MAX_SUBNORMALmm,
                          Double_MAX_SUBNORMAL,
                          DoubleConsts.MIN_NORMAL];
    var someTestCases = [1.0,
                         2.0,
                         3.0,
                         Math.PI,
                         Double_MAX_VALUEmm,
                         Number.MAX_VALUE];
    
    var oneMultiplyScalingFactors = [DoubleConsts.MIN_EXPONENT,
                                     DoubleConsts.MIN_EXPONENT+1,
                                     -3,
                                     -2,
                                     -1,
                                     0,
                                     1,
                                     2,
                                     3,
                                     DoubleConsts.MAX_EXPONENT-1,
                                     DoubleConsts.MAX_EXPONENT];
    var manyScalingFactors = ["-2147483648", // Integer.MIN_VALUE
                              "-2147483647",
                              -MAX_SCALE-1,
                              -MAX_SCALE,
                              -MAX_SCALE+1,
                              2*DoubleConsts.MIN_EXPONENT-1,
                              2*DoubleConsts.MIN_EXPONENT,
                              2*DoubleConsts.MIN_EXPONENT+1,
                              -2,
                              -1,
                              0,
                              1,
                              2,
                              DoubleConsts.MAX_EXPONENT-1,
                              DoubleConsts.MAX_EXPONENT,
                              DoubleConsts.MAX_EXPONENT+1,
                              2*DoubleConsts.MAX_EXPONENT-1,
                              2*DoubleConsts.MAX_EXPONENT,
                              2*DoubleConsts.MAX_EXPONENT+1,
                              MAX_SCALE-1,
                              MAX_SCALE,
                              MAX_SCALE+1,
                              "2147483646",
                              "2147483647"]; // Integer.MAX_VALUE

    it("should handle identity cases", function() {
        for(var i = 0; i < manyScalingFactors.length; i++) {
            expect(util.scalb(NaN, manyScalingFactors[i])).toBeNaN();
            expect(util.scalb(-NaN, manyScalingFactors[i])).toBeNaN();
            for(var j = 0; j < identityCases.length; j++) {
                scalbTest(identityCases[j], manyScalingFactors[i], identityCases[j]);
            }
        }
    });
    it("should handle cases where result is 0.0 or infinity due to magnitude",
       function() {
           for(var i=0; i<someTestCases.length; i++) {
               for(var j=0; j<manyScalingFactors.length; j++) {
                   var scaleFactor = manyScalingFactors[j];
                   if (util.abs(scaleFactor) >= MAX_SCALE) {
                       var val = someTestCases[i];
                       var result = (scaleFactor > 0 ?
                                     (val > 0 ? Infinity : -Infinity)
                                     : (val > 0 ? 0.0 : -0.0));
                       scalbTest(val, scaleFactor, result);
                   }
               }
           }
       });
    it("should handle cases done with one floating-point multiply",
       function() {
           for(var i=0; i<someTestCases.length; i++) {
               for(var j=0; j<oneMultiplyScalingFactors.length; j++) {
                   var scaleFactor = oneMultiplyScalingFactors[j];
                   var val = someTestCases[i];
                   var result = val*util.powerOfTwoD(scaleFactor);
                   scalbTest(val, scaleFactor, result);
               }
           }
       });
});


src/util.js
/**
 * Helper functions used in numerical routines.
 * 
 * @author Alex Nelson
 * @email pqnelson@gmail.com
 * @date 1 October 2014
 */
var exports = module.exports = {};
var DoubleConsts = require('./DoubleConsts.js');

function sgn(x) {
    if (x < 0) return -1;
    if (x > 0) return +1;
    if (x===0) return 0;
    return NaN;
}


/*********************
 * Utility functions *
 *********************/

var isLittleEndian = ((new Uint32Array((new Uint8Array([1,2,3,4])).buffer))[0] === 0x04030201);
var isBigEndian = !isLittleEndian;

function computeEpsilon() {
    if (Number.EPSILON) return Number.EPSILON;
    var k = 1.0;
    while (1.0 !== 1.0+k) {
        k = k*0.5;
    }
    return k;
}

/** Machine epsilon, either given as Number.epsilon or computed from
 *  scratch
 *  @constant
 *  @type {number}
 *  @default
 */
const machineEpsilon = computeEpsilon();
/** The squareroot of machine epsilon
 *  @constant
 *  @type {number}
 *  @default
 */
const sqrtEpsilon = Math.sqrt(machineEpsilon);

/**
 * @todo fix getNumberParts and fromNumberParts to work with Big &
 *       Little Endian numbers
 * taken from @link {http://stackoverflow.com/a/17156580/1296973}
 */
function getNumberParts(x) {
    var float = new Float64Array(1);

    float[0] = x;

    var sign = float.buffer[7] >> 7,
        exponent = ((float.buffer[7] & 0x7f) << 4 | float.buffer[6] >> 4)
                    - DoubleConsts.EXP_BIAS;

    float.buffer[7] = 0x3f;
    float.buffer[6] |= 0xf0;

    return {
        sign: sign,
        exponent: exponent,
        signficand: float[0],
    }
}

function setExponent(f, exponent) {
    var e = exponent+DoubleConsts.EXP_BIAS;
    f.buffer[7] |= (e >> 4);
    f.buffer[6] |= (e & 0x0F) << 4;
    return f;
}


function toggleSignBit(floatArray, isNegative) {
    if (isNegative) {
        floatArray.buffer[7] |= 0x80;
    }
    return floatArray;
}

function clearExponent(f) {
    f.buffer[7] = 0;
    f.buffer[6] &= 0x0f;
    return f;
}

function fromNumberParts(sign, m, exponent) {
    var isNegative = (sign === 1);
    if (exponent > DoubleConsts.MAX_EXPONENT)
        return (isNegative? -Infinity : Infinity);
    
    var f = new Float64Array(1);
    f[0] = m;
    f = clearExponent(f);
    f = toggleSignBit(f, isNegative);
    
    f = setExponent(f, exponent);
    return f[0];
}

function copySign(magnitude, sign) {
    if (sgn(magnitude)===sgn(sign)) return magnitude;

    var f = getNumberParts(magnitude);
    if (f.sign === 0)
        return fromNumberParts(1, f.significand, f.exponent);
    if (f.sign === 1)
        return fromNumberParts(0, f.significand, f.exponent);
}

function getExponent(x) {
    return getNumberParts(x).exponent;
}

function abs(x) {
    if (x < 0) return -x;
    return x;
}

/**
 * Return {@code d} &times; 2<sup>{@code scaleFactor}</sup> rounded
 * as if performed by a single correctly rounded floating-point multiply
 * to a member of the double value set.
 *
 * If the exponent of the result is between -1023 and 1024, the
 * answer is calculated exactly.  If the exponent of the result
 * would be larger than {@code Double.MAX_EXPONENT}, an
 * infinity is returned.  Note that if the result is subnormal,
 * precision may be lost; that is, when {@code scalb(x, n)}
 * is subnormal, {@code scalb(scalb(x, n), -n)} may not equal
 * <i>x</i>.  When the result is non-NaN, the result has the same
 * sign as {@code d}.
 *
 * <p>Special cases:
 * <ul>
 * <li> If the first argument is NaN, NaN is returned.
 * <li> If the first argument is infinite, then an infinity of the same
 * sign is returned. 
 * <li> If the first argument is zero, then a zero of the same
 * sign is returned.
 * </ul>
 * @param d number to be scaled by a power of two.
 * @param scaleFactor power of 2 used to scale {@code d}
 * @return {@code d} &times; 2<sup>{@code scaleFactor}</sup>
 */
// Consider using the http://grepcode.com/file/repo1.maven.org/maven2/org.apache.commons/commons-math/2.2/org/apache/commons/math/util/FastMath.java#FastMath.scalb%28float%2Cint%29
function scalb(d, scaleFactor) {
    if (isNaN(d) || d == 0 || !isFinite(d))
        return d;
    if (scaleFactor < -2098)
        return (d > 0 ? 0.0 : -0.0);
    if (scaleFactor > 2097)
        return (d > 0 ? Number.POSITIVE_INFINITY : Number.NEGATIVE_INFINITY);
    var dNumberParts = getNumberParts(d);
    return fromNumberParts(dNumberParts.sign,
                           dNumberParts.significand,
                           dNumberParts.exponent+scaleFactor);
}

/**
 * Uses Knuth 2.2's method of comparing floating point numbers.
 *
 * If (x < y),    return -1
 *    (x > y),    return +1
 *    (x "eq" y), return  0
 */
function floatCompare(x, y, eps) {
    if (!eps) eps = sqrtEpsilon;
    var exponent = getExponent(abs(x) > abs(y) ? x : y);
    var delta = scalb(eps, exponent);
    var diff = x - y;
    if (diff > delta) return 1;
    if (diff < -delta) return -1;
    return 0;
}

/**
 * @param x, y
 * @returns true if they are "equal", false otherwise
 */
function floatEquals(x, y, eps) {
    return (floatCompare(x, y, eps) === 0);
}

/**
 * Horner's method will take an array of coefficients, and return a
 * function which evaluates the polynomial at the given point.
 */
function horner(coefficients) {
    var n = coefficients.length - 1;

    var fn = function(x) {
        var ret = coefficients[n];
        for(var k = n-1; k > -1; k--) {
            ret = coefficients[k] + x*ret;
        }
        return ret;
    }

    return fn;
}

/**
 * Returns 2^n without doing any multiplication.
 * @param n, some integer
 */
function powerOfTwoD(n) {
    return fromNumberParts(0, 1, n);
}
exports.powerOfTwoD = powerOfTwoD;
exports.isLittleEndian = isLittleEndian;
exports.isBigEndian = isBigEndian;
exports.sgn = sgn;
exports.sqrtEpsilon = sqrtEpsilon;
exports.machineEpsilon = machineEpsilon;
exports.horner = horner;
exports.abs = abs;
exports.getNumberParts = getNumberParts;
exports.fromNumberParts = fromNumberParts;
exports.copySign = copySign;
exports.scalb = scalb;
exports.floatCompare = floatCompare;
exports.floatEquals = floatEquals;


src/rootFinding.js
/**
 * Root finding methods for functions of a single variable.
 *
 * @author Alex Nelson
 * @email pqnelson@gmail.com
 * @date 29 September 2014
 */

var exports = module.exports = {};

require('./util');

/**
 * My todo list...
 * @todo add unit tests!
 * @todo make sure Newton's method won't cause problems with the dy/y code
 */


/**********************************
 * Root finding algorithms        *
 **********************************/
var rootFinding = {};

function bisection(fn, a, b) {
    if (sgn(fn(a)) === sgn(fn(b)))
        return undefined;

    var midpoint;
    var n;
    for(n = 0; n < 53; n++) {
        midpoint = (a + b)*0.5;
        if(fn(midpoint)===0.0 || (b - a)*0.5 < tolerance)
            return midpoint;
        else if (sgn(fn(a))===sgn(fn(midpoint)))
            a = midpoint;
        else
            b = midpoint;
    }
    return midpoint;
}

function secantMethod(fn, a, b) {
    var tmp, x, xPrime, y, yPrime;
    x = a;
    xPrime = (b ? b : a + sqrtEpsilon);
    y = fn(x);
    yPrime = fn(xPrime);
    for(var n = 0; n < 13; n++) {
        if (floatEquals(y, 0.0))
            return x;
        // avoid roundoff errors with this approach
        tmp = (x*yPrime - xPrime*y)/(yPrime - y);
        y = yPrime;
        x = xPrime;
        xPrime = tmp;
        yPrime = fn(xPrime);
    }
    return xPrime;
}

function newtonRootMethod(fn, df, initialGuess) {
    var tmp, y, dy, x;
    x = initialGuess;
    for(var n = 0; n < 9; n++) {
        dy = df(x);
        y = fn(x);
        // @todo should check that dy/y isn't going to cause problems
        x = x - (y/dy);
    }
    return x;
}

rootFinding.bisection = bisection;
rootFinding.secant = secantMethod;
rootFinding.newton = newtonRootMethod;

exports.bisection = bisection;
exports.newton = newtonRootMethod;
exports.secant = secantMethod;
exports.root = rootFinding;


src/DoubleConsts.js
/**
 * Various constants related to IEEE 64-bit double 
 * precision. Inspired/borrowed from Java's sun.misc.DoubleConsts
 * 
 * @author Alex Nelson
 * @email pqnelson@gmail.com
 * @date 11 October 2014
 */

var exports = module.exports = {};

/**
 * A constant holding the smallest positive normal value of type double,
 * 2^-1022. It is equal to the value returned by
 * Double.longBitsToDouble(0x0010000000000000L). 
 */
const MIN_NORMAL = 2.2250738585072014E-308;
/**
 * Specifies the smallest exponent for a double, -1022, in hexadecimal.
 */
const MIN_EXPONENT = -0x3FE;
/**
 * Specifies the largest exponent for a double, 1023, in hexadecimal.
 */
const MAX_EXPONENT = 0x3FF;
/**
 * Specifies the bias for the exponent, 1023, in hexadecimal.
 */
const EXP_BIAS = 0x3FF;
/**
 * Specifies the bits in the Mantissa, 53 for a Double, in hexadecimal.
 */
const SIGNIFICAND_WIDTH = 0x35;
/**
 * Specifies bits where the exponent "lives" in a Double, in hexadecimal.
 * @todo make this endian-independent.
 */
const EXP_BIT_MASK = "0x7ff0000000000000";
const SIGN_BIT_MASK = "0x8000000000000000";
const SIGNIF_BIT_MASK = "0x000FFFFFFFFFFFFF";
const MAX_VALUE = 1.7976931348623157E308;
const MIN_VALUE = 4.9E-324;
const MIN_SUB_EXPONENT = MIN_EXPONENT - (SIGNIFICAND_WIDTH - 1);

exports.EXP_BIAS = EXP_BIAS;
exports.EXP_BIT_MASK = EXP_BIT_MASK;

exports.MAX_EXPONENT = MAX_EXPONENT;
exports.MAX_VALUE = MAX_VALUE;

exports.MIN_EXPONENT = MIN_EXPONENT;
exports.MIN_NORMAL = MIN_NORMAL;
exports.MIN_SUB_EXPONENT = MIN_SUB_EXPONENT;
exports.MIN_VALUE = MIN_VALUE;

exports.NaN = NaN;
exports.NEGATIVE_INFINITY = -Infinity;
exports.POSITIVE_INFINITY = Infinity;

exports.SIGN_BIT_MASK = SIGN_BIT_MASK;
exports.SIGNIF_BIT_MASK = SIGNIF_BIT_MASK;
exports.SIGNIFICAND_WIDTH = SIGNIFICAND_WIDTH;
